### PR TITLE
Add sort to product page

### PR DESCRIPTION
**Pull request by kul3r4**
Solves #829

### CODE REVIEW

**Comment by kul3r4** [comment]
@sebastianbenz 
I managed to remove the rendering of the products server side and do that with amp-list/amp-bind.
This also required to change how the search works and given the search is used in the product sample as well, for now I duplicated the endpoints to live the product page as it is.

If you are ok with the approach then I will move the Product Page to the same code (html and backend)

Find a preview here: [https://amp-by-example-staging.appspot.com/samples_templates/product_browse_page/preview/](https://amp-by-example-staging.appspot.com/samples_templates/product_browse_page/preview/).

If you also try on mobile, the first time you load the page the show-more button appears, then after a search, even if the content of amp-list changes, the show more button stay, this is likely a bug and I will follow up separately.

**Comment by kul3r4** [comment]
@sebastianbenz I added the support for keeping the search when sorting. PTAL

**Comment by kul3r4** [comment]
@sebastianbenz I changed the layout of the page to be a list instead of a carousel. PTAL

**Comment by sebastianbenz** [comment]
Few comments:

* could we add a reset button instead of an `all` list item?
* can you make the product list show up in a grid as before? This looks weird on desktop with a single column
* can you move the sorting input into the filter bar?
* suggestion for sorting input: change label to 'sort by' and values to 'price lowest', 'price highest'

**Comment by kul3r4** [comment]
Hi,
thanks for reviewing, my answers in line:
could we add a reset button instead of an all list item?
I am adding it for the color, yes
can you make the product list show up in a grid as before? This looks weird on desktop with a single column
This was a suggestion from @ericlindley-g, you can find the discussion in #829 
can you move the sorting input into the filter bar?
will do
suggestion for sorting input: change label to 'sort by' and values to 'price lowest', 'price highest'

**Comment by sebastianbenz** [comment]
Hm. I'd still prefer the grid view. I think the sorting will still be obvious (we could add more products to make this more explicit). Mobil will still be a single column.

**Comment by kul3r4** [comment]
@sebastianbenz you can take a look at existing websites like:
[ebay](https://www.ebay.co.uk/sch/i.html?_from=R40&_trksid=p2380057.m570.l1313.TR0.TRC0.H0.Xcamera.TRS0&_nkw=camera&_sacat=0)
[bh](https://www.bhphotovideo.com/c/search?Ns=p_PRICE_2%7c0&ci=16158&N=4288586281)
there you can find the pattern of using a list for the category page

**Comment by sebastianbenz** [comment]
Yes, but then the product list items need to be responsive. The whitespace
too the right is too ugly imho.

**Comment by kul3r4** [comment]
I agree, we need to find some the content to add to the right

**Comment by kul3r4** [comment]
@sebastianbenz @ericlindley-g I think it would be nice to recap where we are with this PR to decide next steps and see if we can break down additional features:
[staging](https://amp-by-example-staging.appspot.com/samples_templates/product_browse_page/preview/) contains a version where: 
- sorting server side is implemented
- desktop view has color filters on the left and mobile has color filter in a ligthbox (this is using amp-list and amp-bind together as well)

Outstanding problems:
- amp-list requires a fixed height and we are using amp-list for showing results . As the content of the amp-list changes based on the query, if the query returns a small result (like only one product) then the space used by the amp-list is too high but there is no way of resizing
- should we agree on what is the best design for the list of products? I think list is a cleaner way of showing results in comparison to the grid, this is also supported by existing samples I pointed above
- if we keep the list layout, I would need to add something on the right of the product to not leave the blank space on desktop. I suggest an additional column with some product info, WDYT?
- at the moment it's possible to reset the color filter but not the search (you would need to reload). Any recommendation here?

Thanks you very much for your support and review, this sam

**Comment by ericlindley-g** [comment]
Sorting looks amazing! It's great to see that this all comes together so well

w/r/t grid v. list, I think we could make either work if we make some changes to the design.

Grid view: needs more products to be comprehensible: when it's just a single row of items it looks more like a "featured products" area. I'm not sure this will be enough to make the interaction understandable, though. Also might help to put the items on "cards", but it's hard to be sure.

List view: This felt much more "clear" to me, and also makes sorting more straightforward, because the sort order is only top to bottom (as opposed to the case in rows, where the sort is left to right, and top to bottom). However, it needs a visual design that fills up more of the horizontal space to look OK on desktop. This could mean adding more information, or spreading the existing information across a wider card.

In the end, I recommended the list view because that seemed like the most straightforward way to make this interaction clear to the user, but if others feel strongly, I could see either interaction being clear with enough design thought.

/cc @spacedino for any additional thoughts

**Comment by sebastianbenz** [comment]
I'm happy with both grid or list, the only thing we need to fix for the list is the horizontal space. Grid seems like a more common implementation. We need to add more products anyway if we want to demo pagination as well...

**Comment by ericlindley-g** [comment]
Sounds good — I'm hoping to include a GIF of this interaction in the amp-bind blog post coming soon. What do you think about merging this version for now so we can direct people to it in the short term? Then we can settle on either grid or list (or both, if we want a toggle between them like [this page](https://www.usnews.com/best-colleges/rankings/national-universities?_mode=list) on desktop)

Another option is adding more products and switching back to grid, but I'm not 100% sure that will fix the clarity issue of the interaction.

**Comment by sebastianbenz** [comment]
+1 for adding more products and switching back to grid -> on mobile you still get a single column you can use for your GIF.

@kul3r4  might be worth checking out [this article](http://heydonworks.com/practical_aria_examples/#toolbar-widget) to see if we can do anything to improve accessibility of the sorting/filter buttons.

**Comment by kul3r4** [comment]
Just deployed on [staging](https://amp-by-example-staging.appspot.com/samples_templates/product_browse_page/preview/) a new version with 2 more items and put back the grid for desktop. @sebastianbenz WDYT?

**Comment by sebastianbenz** [comment]
Looks great on desktop! Seems like there are still a few quirks on mobile:

1. Some results don't render correctly in the list

screen shot 2017-07-05 at 11 24 43 am

2. Can you move the close button in the filter lightbox to the bottom?

screen shot 2017-07-05 at 11 25 20 am

3. Suggestion: on mobile can you put the filter and sort input in the same line?

**Comment by kul3r4** [comment]
@sebastianbenz Thanks! just fixed those and deployed to staging.

File: src/50_Samples_%26_Templates/Product_Browse_Page.html [hunk 1]
```diff
@@ -1,218 +1,415 @@
...
-        }
-    }
-  </script>
-</amp-analytics>
-</body>
+         </script>
+      </amp-state>
+      <amp-lightbox id="filter-lightbox" layout="nodisplay">
+         <div>
+            <h3 class="ml1">Filter results</h3>
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Browse_Page.html]: s/ml1/m1 mt2/

**Comment by sebastianbenz** [comment]
great! one small style nit. 

Is it possible to close the Lightbox when you select a filter option?

**Comment by kul3r4** [comment]
@sebastianbenz I removed the close button and now it closes when selecting a color.

**Comment by sebastianbenz** [comment]
@kul3r4 sorry for not being precise. We should leave the close button as well -> in case someone doesn't want to select a filter.

**Comment by kul3r4** [comment]
@sebastianbenz you are right, I added it back

**Comment by sebastianbenz** [comment]
Great!

**Pull request merged**